### PR TITLE
[GSoC] Kotlin migration Connection.java

### PR DESCRIPTION
**Pull request by divyansh-dxn**
## Purpose / Description
This PR is a part of `GSoC 2022 - AsyncTask to Coroutines`. This class uses AsyncTask, which needs to be migrated to Kotlin first before using Coroutines.
Kotlin conversion : Connection.java -> Connection.kt

## Approach
- Convert the JAVA file to Kotlin
- Do some small changes manually

## How Has This Been Tested?
Runs as expected on `Realme 6i API 30` and `Pixel 4 Emulator API 31`
Passes all written tests.

https://user-images.githubusercontent.com/69595691/173822488-5faedc58-29c7-46e6-a59d-b200ebb1e724.mp4

https://user-images.githubusercontent.com/69595691/173822466-caae38e3-5744-4d12-9cc0-ad86723af249.mp4

_entering credentials part has been removed_

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [ ] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison — COMMENTED** [review]
```

> Task :AnkiDroid:compilePlayDebugAndroidTestJavaWithJavac FAILED
/Users/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/libanki/HttpTest.java:68: error: cannot find symbol
        if (!Connection.isOnline()) {
                       ^
  symbol:   method isOnline()
  location: class Connection
1 error
```

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+                            }
+                        }
+                        ConflictResolution.FULL_DOWNLOAD -> {
+                            Timber.i("Sync - fullsync - download collection")
+                            publishProgress(R.string.sync_downloading_message)
+                            val ret = fullSyncServer.download()
+//                            if (ret == null) {
+//                                Timber.w("Sync - fullsync - unknown error")
+//                                return returnGenericError(data)
+//                            }
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:310]: what is the reason for commenting this code?

Comment (divyansh-dxn) [reply AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:310]: Nothing, I just forgot to remove it. Earlier based on the nullability of `ret`, the message was returned but after conversion, it is of non-null type. I'll just remove it.

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 2]
```diff
@@ -16,666 +16,600 @@
...
-                    if (downloadedCount == 0 && uploadedCount == 0) {
-                        mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage();
+                    val downloadedCount = mediaClient.getDownloadCount()
+                    val uploadedCount = mediaClient.getUploadCount()
+                    mediaError = if (downloadedCount == 0 && uploadedCount == 0) {
+                        """
+     ${AnkiDroidApp.getAppResources().getString(R.string.sync_media_error)}
+     
+     ${e.localizedMessage}
+                        """.trimIndent()
```
Comment (divyansh-dxn) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:400]: After conversion, this part is dangling with spaces and quotes, should I mark it for cleanup or just correct it in the 2nd commit?

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 3]
```diff
@@ -16,666 +16,600 @@
...
+     ${
+                        AnkiDroidApp.getAppResources().getString(
+                            R.string.sync_media_partial_updated,
+                            downloadedCount,
+                            uploadedCount
+                        )
+                        }
+     
+     ${e.localizedMessage}
+                        """.trimIndent()
```
Comment (divyansh-dxn) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:412]: same

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt (hunk 2 above)
Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:400]: if it's a small change, correct it right away

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 4]
```diff
@@ -16,666 +16,589 @@
...
+import com.ichi2.utils.KotlinCleanup
+import com.ichi2.utils.Permissions.canUseWakeLock
+import okhttp3.Response
+import timber.log.Timber
+import java.io.IOException
+import java.util.*
+
+@Suppress("DEPRECATION") // #7108: AsyncTask
+@KotlinCleanup("Simplify null comparison, !! -> ?.")
+class Connection : BaseAsyncTask<Connection.Payload, Any, Connection.Payload>() {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:56]: KotlinCleanup: IDE-lint

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 5]
```diff
@@ -16,666 +16,589 @@
...
+        var taskType = 0
+        var resultType: ConnectionResultType? = null
+        var result: Array<Any?> = arrayOf()
+        @JvmField
+        var success = true
+        var returnType = 0
+        var exception: Exception? = null
+        var message: String? = null
+        var col: Collection? = null
+        override fun toString(): String {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:492]: KotlinCleanup: use formatted string

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 6]
```diff
@@ -16,666 +16,589 @@
...
-    }
-
-
-    private Payload doInBackgroundLogin(Payload data) {
-        String username = (String) data.data[0];
-        String password = (String) data.data[1];
-        HostNum hostNum = (HostNum) data.data[2];
-        RemoteServer server = new RemoteServer(this, null, hostNum);
-        Response ret;
+    private fun doInBackgroundLogin(data: Payload): Payload {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:132]: KotlinCleanup: Scoped function

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 7]
```diff
@@ -16,666 +16,589 @@
...
+        val conflictResolution = data.data[2] as ConflictResolution?
         // A number AnkiWeb told us to send back. Probably to choose the best server for the user
-        HostNum hostNum = (HostNum) data.data[3];
+        val hostNum = data.data[3] as HostNum
         // Use safe version that catches exceptions so that full sync is still possible
-        Collection col = CollectionHelper.getInstance().getColSafe(AnkiDroidApp.getInstance());
-
-        boolean colCorruptFullSync = false;
+        val col = CollectionHelper.getInstance().getColSafe(AnkiDroidApp.getInstance())
+        var colCorruptFullSync = false
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:249]: KotlinCleanup: Make colCorruptFullSync a val

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 8]
```diff
@@ -16,666 +16,589 @@
...
-
-        public Payload(@NonNull Object[] data) {
-            this.data = data;
-            success = true;
+        /**
+         * Add generic error value to the payload
+         * @param data Some payload that should be transformed
+         * @return the original payload
+         */
+        private fun returnGenericError(data: Payload): Payload {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:557]: KotlinCleanup: remove `return` from the method name
KotlinCleanup: scoped function

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 9]
```diff
@@ -16,666 +16,589 @@
...
-    }
-
     /**
      * In the payload, success means that the sync did occur correctly and that a change did occur.
-     * So success can be false without error, if no change occurred at all.*/
-    private Payload doInBackgroundSync(Payload data) {
-        sIsCancellable = true;
-        Timber.d("doInBackgroundSync()");
+     * So success can be false without error, if no change occurred at all. */
+    private fun doInBackgroundSync(data: Payload): Payload {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:233]: KotlinCleanup: scoped function

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 10]
```diff
@@ -16,666 +16,589 @@
...
+        var isCancelled = false
+            private set
+
+        @get:Synchronized
+        var isCancellable = false
+            private set
+        var allowLoginSyncOnNoConnection = false
+
+        // #7108: AsyncTask
+        private fun launchConnectionTask(listener: TaskListener, data: Payload): Connection? {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:522]: KotlinCleanup: Scoped function

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 11]
```diff
@@ -16,666 +16,589 @@
...
 
+        /* NetworkInfo is deprecated in API 29 so we have to check separately for higher API Levels */
+        val isOnline: Boolean
+            get() {
+                if (allowLoginSyncOnNoConnection) {
+                    return true
+                }
+                val cm = AnkiDroidApp.getInstance().applicationContext
+                    .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+                /* NetworkInfo is deprecated in API 29 so we have to check separately for higher API Levels */return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:572]: return if... etc should be on the next line

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 12]
```diff
@@ -16,666 +16,589 @@
...
+import com.ichi2.utils.Permissions.canUseWakeLock
+import okhttp3.Response
+import timber.log.Timber
+import java.io.IOException
+import java.util.*
+
+@Suppress("DEPRECATION") // #7108: AsyncTask
+@KotlinCleanup("Simplify null comparison, !! -> ?.")
+class Connection : BaseAsyncTask<Connection.Payload, Any, Connection.Payload>() {
+    private var mListener: TaskListener? = null
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:57]: KotlinCleanup: lateinit

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 13]
```diff
@@ -16,666 +16,589 @@
...
+import com.ichi2.libanki.sync.HttpSyncer
+import com.ichi2.libanki.sync.MediaSyncer
+import com.ichi2.libanki.sync.RemoteMediaServer
+import com.ichi2.libanki.sync.RemoteServer
+import com.ichi2.libanki.sync.Syncer
+import com.ichi2.libanki.sync.Syncer.ConnectionResultType
+import com.ichi2.utils.JSONException
+import com.ichi2.utils.JSONObject
+import com.ichi2.utils.KotlinCleanup
+import com.ichi2.utils.Permissions.canUseWakeLock
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:48]: Remove this import

**Review by BrayanDSO — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.kt [hunk 14]
```diff
@@ -16,666 +16,603 @@
...
+import com.ichi2.libanki.sync.Syncer
+import com.ichi2.libanki.sync.Syncer.ConnectionResultType.*
+import com.ichi2.utils.JSONException
+import com.ichi2.utils.JSONObject
+import com.ichi2.utils.KotlinCleanup
+import com.ichi2.utils.Permissions
+import okhttp3.Response
+import timber.log.Timber
+import java.io.IOException
+import java.util.*
```
Comment (divyansh-dxn) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.kt:53]: Avoid wildcard imports

**Review by BrayanDSO — APPROVED** [review]
LGTM

**Review by Akshay0701 — APPROVED** [review]

**Comment by Akshay0701** [comment]
Thanks @divyansh-dxn

**Review by david-allison — APPROVED** [review]

**Pull request merged**